contracts/tests/mcms/ManyChainMultiSigExecute.spec.ts

```diff
@@ -0,0 +1,437 @@
+import { toNano, beginCell } from '@ton/core'
+import '@ton/test-utils'
+import { MCMSBaseSetRootAndExecuteTestSetup, MCMSTestCode } from './ManyChainMultiSigBaseTest'
+import * as mcms from '../../wrappers/mcms/MCMS'
+import { asSnakeData } from '../../src/utils'
+import { ERROR_UNAUTHORIZED_SIGNER } from '../../wrappers/libraries/ocr/ExitCodes'
+
+describe('MCMS - ManyChainMultiSigExecuteTest', () => {
+  let baseTest: MCMSBaseSetRootAndExecuteTestSetup
+  let code: MCMSTestCode
+
+  beforeAll(async () => {
+    code = await MCMSBaseSetRootAndExecuteTestSetup.compileContracts()
+  })
+
+  beforeEach(async () => {
+    baseTest = new MCMSBaseSetRootAndExecuteTestSetup()
+    baseTest.code = code
+    await baseTest.setupForSetRootAndExecute('test-execute')
+    await baseTest.setInitialRoot()
+  })
+
+  it('should revert when post-op count reached', async () => {
+    // Fund for value op
+    await baseTest.bind.mcms.sendInternal(
+      baseTest.acc.deployer.getSender(),
+      toNano('10'),
+      mcms.builder.message.in.topUp.encode({
+        queryId: BigInt(1),
+      }),
+    )
+
+    // Execute all operations up to the post-op count limit to simulate setOpCount // TODO this could be hidden behind some helper. It is not the focus of this test
+    await baseTest.advanceOpcodeTo(MCMSBaseSetRootAndExecuteTestSetup.OPS_NUM)
+
+    // Verify we've reached the post-op count
+    const currentOpCount = await baseTest.bind.mcms.getOpCount()
+    expect(Number(currentOpCount)).toEqual(MCMSBaseSetRootAndExecuteTestSetup.OPS_NUM)
+
+    // Now try to execute one more operation - should fail with PostOpCountReached
+    // Use any operation and proof - they won't even be checked
+    const fakeOp = baseTest.testOps[0]
+    const fakeProof = asSnakeData<bigint>([], (v) => beginCell().storeUint(v, 256))
+
+    const executeBody = mcms.builder.message.in.execute.encode({
+      queryId: 999n,
+      op: mcms.builder.data.op.encode(fakeOp),
+      proof: fakeProof,
+    })
+
+    const result = await baseTest.bind.mcms.sendInternal(
+      baseTest.acc.deployer.getSender(),
+      toNano('1'),
+      executeBody,
+    )
+
+    expect(result.transactions).toHaveTransaction({
+      from: baseTest.acc.deployer.address,
+      to: baseTest.bind.mcms.address,
+      success: false,
+      exitCode: mcms.Error.POST_OP_COUNT_REACHED,
+    })
+  })
+
+  it('should revert on bad proof', async () => {
+    // Modify the first op by incrementing value
+    const modifiedOp = { ...baseTest.testOps[0] }
+    modifiedOp.value = modifiedOp.value + 1n
+
+    // Try with empty proof first
+    const emptyProof = asSnakeData<bigint>([], (v) => beginCell().storeUint(v, 256))
```

Comment:
We should move this and few other encodings to contract bindings (data CellCodec), as they are currently repeated many times across tests.